### ISSUE TITLE
[BUG]: We are using PayPal Hosted Fields from library. The order is not approved when we call submit function from hostedFields, we are getting response with orderId but it is not approved it's status is still Created.

### ISSUE THREAD

**Issue by codio-lab** [paypal/react-paypal-js#254]
### Is there an existing issue for this?

- [X] I have searched the existing issues.

### 🐞 Describe the Bug

we have implemented according to this storybook component.
https://paypal.github.io/react-paypal-js/?path=/docs/paypal-paypalhostedfields--default
we are facing an issue request urgent response.

### 🌍 Environment

```markdown
| Software         | Version(s) |
| ---------------- | ---------- |
| react-paypal-js  |            |
| Browser          |            |
| Operating System |            |
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by xjamundx** [comment]
In the logged output are you seeing any error messages? If so could you
share those here? Also any snippets of your code?

Can you also confirm you are using a valid client id and client token?

Do you know id this issue is happening in the sandbox environment or for
production?

**Comment by codio-lab** [comment]
Hi, @xjamundx
I have attached a screencast of the entire flow. [https://share.vidyard.com/watch/g5QzgtVKoeQH9yWfZTqzm5?]
we are not getting any error from submit method provided by hosted fields instance. Getting orderId in the "then" block.
checked order with the postman and find that is still in 'CREATED' status. the failed API mentioned screencast is from our own server responding error because the order is not approved.

To make sure we checked again and confirmed that are using the correct client_id and client token.
we are using the local environment. it will be deployed in production once done.

**Comment by codio-lab** [comment]
> `[](url)`

**Comment by codio-lab** [comment]
@xjamundx  can you please help us to understand how can we achieve this feature implementation?

**Comment by diegoag64** [comment]
I am having a similar issue. I have been working on implementing paypal with a react project and the it seems like paypal changed something on their end that makes the documentation for this library not work anymore.

Initially I tried my best to debug the issue, but without any solution I emailed paypal developer support and they told me that to process direct credit card transactions, the payment_source has to be sent when creating the order. The confusing part is that they dont mention that at all in their docs, but anyways I tried it, and after debugging other issues it finally worked with Postman.

Now my issue is that with the information provided in the README, I dont understand how to pass the payment_source to my server create order implementation. 

I hope this helps everyone having this issue, I am still emailing back and forth with paypal to try to resolve my issue, since its hard to make them understand what is really happening and they dont seem to acknowledge this library at all.

**Comment by codio-lab** [comment]
@diegoag64 can you please provide me a sample request body for creating an order API where you are passing 'payment_source' ?

**Comment by diegoag64** [comment]
@codio-lab it would look like this:

`{
    "intent": "CAPTURE",
        "purchase_units": [
          {
            "amount": {
              "currency_code": "USD",
              "value": "50000"
            }
          }
        ],
        "payment_source": {
            "card": {
                "number": "4032031108790766",
                "expiry": "2024-06",
                "security_code": "336",
                "name": "John Doe"
            }
        }
}`

But of course they couldnt give me all the information I needed and I realized I also need to add a "Paypal-Request-Id" header, which contains a random key to reference the transaction later, without that it doesnt work.

**Comment by diegoag64** [comment]
@xjamundx anyone paying attention to this issue? My project has been stuck for a week because this is not working and I was hoping paypal would help but I feel lost here.

Contacted paypal support and they just said "Ask on github" so here I am...

**Comment by xjamundx** [comment]
@diegoag64 Ahh sorry I saw earlier you said you were working with support on this issue and I was hoping things were going well. I'll get someone from the team to look at this ASAP

**Comment by xjamundx** [comment]
@diegoag64 BTW we're working on a demo codesandbox that does advanced credit card processing with the react sdk (using a nextjs app). It's not something we're promoting quite yet (and still in code review phase), but it might offer some help to see a mostly working example with the server component. That's available at https://codesandbox.io/s/paypal-custom-payment-integration-b9il3r

Trying to find the right person from the SDK teams that knows the ins and outs of this stuff, to make sure to get you the right answer (and fix any issues that may exist in the docs)

**Comment by ankuronweb** [comment]
@diegoag64 By seeing the video you provided looks like we are missing the last step here- 
4. Capture Order- https://developer.paypal.com/docs/checkout/advanced/integrate/#link-captureorder
Please take a look at this sample code on line #32 where you need to call the Orders API with the returned Order id- 
https://codesandbox.io/s/paypal-custom-payment-integration-b9il3r?file=/pages/components/cardForm.js:1039-1123

**Comment by diegoag64** [comment]
@ankuronweb That is where I am confused, I followed the example provided in the documentation with the process of creating the order and then capturing it, but my transaction always stayed as PENDING when I did that.

I contacted paypal support and they told me that the Create Order -> Capture Order flow was only for the paypal buttons, but for hosted fields I just needed to create the order with a payment_source and it would be captured right away.

I replicated what was suggested with Postman and it works fine making the direct request to the orders api, but I dont understand why creating order and capturing is done in one step now but the documentation (and the codesandbox example) does the opposite.

Also I tested the codesandbox example and it doesnt work, seems to be having my same problem now?

**Comment by ankuronweb** [comment]
@diegoag64 I just tested Codesandbox demo. It is working fine. Are you using one of the test credit cards or a real card?
<img width="656" alt="Screen Shot 2022-02-28 at 11 17 14 AM" src="https://user-images.githubusercontent.com/1686024/156046318-a9f6486c-2588-486b-b421-7de799a6f13e.png">

**Comment by diegoag64** [comment]
@ankuronweb Oh yeah sorry I was trying with a random card. Its working for me too with that number. I am trying to fork the sandbox and do some more debugging with the responses that come back, because I did get the same "COMPLETED" response for the order but for me inside of the purchase_units the payment was always stuck on PENDING and the transaction never showed on my sandbox merchant account.

I will mess around with the sandbox, im currently having some issues forking it but should be able to start testing soon.

**Comment by diegoag64** [comment]
@ankuronweb Is it confirmed that that sandbox actually makes transactions that show in the merchant account? I forked it and replaced the client_id and client_secret with mine, transactions seem to be working in the front-end (like it did in my previous implementation), but the transactions are not visible in the activity page of the sandbox merchant account.

**Comment by ankuronweb** [comment]
@diegoag64 Yes, you should be able to see the transaction on your Sandbox Merchant Dashboard once your Orders API call is successful. Just to double check- have you completed the first step here?- 

`1. Enable your account`
https://developer.paypal.com/docs/checkout/advanced/integrate/

Please use your SB account email and complete the flow by filing in dummy business details

**Comment by diegoag64** [comment]
@ankuronweb yes I already completed that step, I am able to see the transactions on the sandbox API Calls page but they dont show up on my merchant's account. If I provide you the Paypal debug ID would you be able to take a look for me?

I just forked the codesandbox code and replaced the client_id and client_secret, everything should work technically.

API Calls Debug IDs:
Order Creation: d3d2646520264
Order Capture: ba200559e3f1a

**Comment by menglander** [comment]
@ankuronweb @xjamundx I'm also having an issue with figuring this out. Do you have a contact for someone that can help with this?

**Comment by xjamundx** [comment]
@menglander Have you contacted merchant support?  Have you found a bug with the SDK or do you need implementation assistance?

@diegoag64 I'm sorry I don't have access to anything support does in terms of verifying your individual account. What I think you're saying is the transactions appear to be working, but aren't showing up on the API calls page on your developer dashboard? Or are the transactions *not* working at all?

**Comment by diegoag64** [comment]
@xjamundx how can I get in direct contact with someone that can help me with this? I contacted merchant dev support and they told me to come here. This should be an easy issue to understand but its becoming so difficult.

The transaction seems to be working fine in your demo, but I cant see it in the Activity page of my sandbox merchant account. I forked the codesandbox and just replaced the client_id and secret_key with mine (third time im mentioning this), entered the test card, transaction seems to work but it never shows on the merchant account.

I feel like im going in circles here trying to explain my issue, in my first comment on this thread I explain exactly what im going through but paypal devs dont seem to understand. How can I be more clear?

**Comment by xjamundx** [comment]
@diegoag64 I had a member of our MTS team review the issue and verify that the transaction ID of the capture debug_id you posted earlier did go through properly (captured with `status:COMPLETED`).  They’ll be reaching out to get more information and make sure you’re able to find these completed transactions in the receiving sandbox account, and help with any other issues.

FYI, I work in developer relations, so I'm mostly going through these github issues looking for common issues that might need documentation improvements 👀  or better code samples. Sorry I wasn't more helpful myself!

**Comment by menglander** [comment]
Thanks. Is what we're implementing widely used among merchants?

We've got 18 accounts with PayPal and really hoping we can get this
resolved and implemented fast.

Thanks again for your help.

**Comment by xjamundx** [comment]
@menglander The React PayPal SDK is super popular, but yeah reach out to support with more details and we'll get it figured out. I obviously don't have any specific information about what you're trying to accomplish at this point.

**Comment by diegoag64** [comment]
@xjamundx Thanks! Yeah I just checked my merchant account and how all the transactions are there and the demo is working fine.

Do you have any idea of why support initially suggested me to include the payment_source on the request when creating the order and that the capture was done automatically? That is what let me into the rabbithole that I was in since none of the documentation mentioned any of that. Maybe they were confused with what I was trying to do?

This is exactly what they sent to me:
_If you wish to fund the transaction from a card, you need to pass the "payment_source" object within the order creation request, not the order capture request.  Once the order is created, you submit the capture request against the order id which is returned within the create order API response._

**Comment by xjamundx** [comment]
@diegoag64 I agree that doesn't sound right, the recommendation I've heard is to use HostedFields which don't require you manually including the `payment_source`. I'm so glad you were able to get it working though and I promise we will improve our documentation in this area.

**Comment by xjamundx** [comment]
Closing this issue for now. To summarize a few things that came up:

1. When sending to/from accounts that are unverified transactions will end up in a pending state and not go through. Make sure to verify those accounts!
2. Use the `<PayPalHostedFieldsProvider>` and `<PayPalHostedField>` components to safely link to credit card related fields (instead of manually populating the `payment_source`
3. We now have a codesandbox link with a full-stack implementation you can look at for inspiration and more insight into what needs to happen on the server to complete a transaction:
 https://codesandbox.io/s/paypal-custom-payment-integration-b9il3r
